### ISSUE TITLE
Fastify - Bad DX error handling

### ISSUE THREAD

**Issue by StarpTech** [connectrpc/connect-es#706]
**Is your feature request related to a problem? Please describe.**
Following this example https://github.com/bufbuild/connect-es-integration/tree/main/fastify any error thrown inside a RPC implementation is not logged to the console. The error is also not propagated to the [Fastify error handler.](https://fastify.dev/docs/latest/Reference/Errors/). It seems the error is never propagated or logged https://github.com/bufbuild/connect-es/blob/main/packages/connect/src/protocol-connect/handler-factory.ts#L248-L265 The dev has no idea why the RPC call failed.

**Describe the solution you'd like**

- Allow to define an error handler __or__
-  Pass the error to the fastify logger. __or__
- Throw the error in the fastify handler and provide a default implementation so we can handle them natively with fastify.

Please also add an "Errors" section to the Node.js docs https://connect.build/docs/node/getting-started

Please specify whether the request is for Connect for Web or Connect for
Node.js.

Connect for Node.js

**Comment by yoitsro** [comment]
This is also affecting us. Just from an expectations perspective, is it likely this will get looked at in the near future or not so much?

**Comment by srikrsna-buf** [comment]
We just released support for [server side interceptors](https://connectrpc.com/docs/node/interceptors), with this you can catch errors that are raised inside the RPC handlers:

```ts
import * as http from "http";
import routes from "./connect";
import { fastifyConnectPlugin } from "@connectrpc/connect-fastify";
import type { Interceptor } from "@connectrpc/connect";

const logger: Interceptor = (next) => async (req) => {
  try {
    return await next(req);
  } catch (e) {
    const ce = ConnectError.from(e);
    console.log(ce.message);
    throw e;
  } 
};

const server = fastify();

await server.register(fastifyConnectPlugin, {
 routes,
 interceptors: [logger],
});

await server.listen({
  host: "localhost",
  port: 8080,
});
```

You can also use the `contextValues` pass the fastify logger and access that from the interceptor.

**Comment by StarpTech** [comment]
Great news! Thank you. Have you also thought about leveraging the existing technology, Fastify, to implement such functionality? Fastify has an extensive hook system. I find it not intuitive when features are re-implemented, although I understand the reason (easier to implement). In the end, all server plugins are abstracted by the connect interface, and you can't really go just the framework way.

**Comment by StarpTech** [comment]
Regarding the issue. As far I understand. Interceptor doesn't solve the issue. Any error in the handler are swallowed https://github.com/connectrpc/connect-es/blob/main/packages/connect-fastify/src/fastify-connect-plugin.ts#L133-L140 and can't be catched by fastify global error handler.

**Comment by srikrsna-buf** [comment]
> Any error in the handler are swallowed https://github.com/connectrpc/connect-es/blob/main/packages/connect-fastify/src/fastify-connect-plugin.ts#L133-L140

This is not related to implementation errors that are thrown in the handlers, this is in situations where the connect itself fails to write for some reason. This is fallback to not miss such errors.

**Comment by StarpTech** [comment]
But the error won't be catchable by fastify global error handler correct?

**Comment by srikrsna-buf** [comment]
> But the error won't be catchable by fastify global error handler correct?

Are you referring to protocol errors or handler errors? 

Connect handles the latter and sends the appropriate error response back to the client. If you want to log them you can use the interceptor to do so. This is equivalent of writing back an error status code in plain fastify.

Protocol errors are unexpected and are currently logged in the snippet you pointed to, if you think there is an idiomatic way of doing that, feel free to open a PR for that.

**Comment by StarpTech** [comment]
Thanks for clarifying! I'd like to come back to the initial issue. Do you see a way to leverage fastify more to avoid thinking in connectrpc and fastify? E.g how can we improve the integration in a way that connectrpc handlers behave the same as fastify handlers?

**Comment by srikrsna-buf** [comment]
> how can we improve the integration in a way that connectrpc handlers behave the same as fastify handlers? That would allow to use fastify as it is with all hooks.

I am not quite sure what you mean here. Can you open a PR with how you think it should look?

**Comment by StarpTech** [comment]
Let me rephrase. Connectrpc register routes through fastify. Because it is an integration for fastify, as a dev, I'd expect I can use fastify as it is. This includes the [hooks](https://fastify.dev/docs/latest/Reference/Hooks/) of the [full request / response life cycle](https://fastify.dev/docs/latest/Reference/Lifecycle/#lifecycle).
